### PR TITLE
[NM] support external location for tables

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Fixes #issue

### CODE REVIEW

File: src/query/service/src/sessions/query_ctx.rs [hunk 1]
```diff
@@ -359,8 +362,14 @@ impl TableContext for QueryContext {
         self.shared.get_query_str()
     }
     // Get the storage data accessor operator from the session manager.
-    fn get_storage_operator(&self) -> Result<Operator> {
-        let operator = self.shared.storage_operator.clone();
+    fn get_storage_operator(&self, sp: Option<StorageParams>) -> Result<Operator> {
```
Comment (Xuanwo) [inline src/query/service/src/sessions/query_ctx.rs:365]: I prefer the following API:

```rust
fn get_internal_storage_operator(&self) -> Operator
fn get_external_storage_operator(&self, sp: StorageParams) -> Result<Operator>
```

File: src/query/storages/fuse/src/io/read/meta_readers.rs [hunk 2]
```diff
@@ -96,17 +115,28 @@ where T: BufReaderProvider + Sync + Send
...
 impl BufReaderProvider for &dyn TableContext {
-    async fn buf_reader(&self, path: &str, len: Option<u64>) -> Result<BufReader<BytesReader>> {
-        let operator = self.get_storage_operator()?;
+    async fn buf_reader(
+        &self,
+        path: &str,
+        len: Option<u64>,
+        sp: Option<StorageParams>,
+    ) -> Result<BufReader<BytesReader>> {
+        let operator = self.get_storage_operator(sp)?;
```
Comment (Xuanwo) [inline src/query/storages/fuse/src/io/read/meta_readers.rs:139]: Cloning an operator is cheap, but building an operator is far more expensive, especially for services like s3.

Try our best to reuse the same operator.

File: src/query/storages/fuse/src/operations/commit.rs [hunk 3]
```diff
@@ -123,37 +124,42 @@ impl FuseTable {
...
-                        break Err(ErrorCode::OCCRetryFailure(format!(
-                            "can not fulfill the tx after retries({} times, {} ms), aborted. table name {}, identity {}",
-                            retry_times,
-                            Instant::now()
-                                .duration_since(backoff.start_time)
-                                .as_millis(),
-                            tbl.table_info.name.as_str(),
-                            tbl.table_info.ident,
-                        )));
+                Err(e) if self::utils::is_error_recoverable(&e, transient) => {
```
Comment (Xuanwo) [inline src/query/storages/fuse/src/operations/commit.rs:127]: OpenDAL has its logic for backoff. And I have a plan to refactor here. So we can return directly here to make things simpler.

File: src/meta/protos/proto/table.proto [hunk 4]
```diff
@@ -124,6 +124,8 @@ message TableMeta {
 
   // Comments of each field.
   repeated string field_comments = 25;
+
+  optional string storage_params = 26;
```
Comment (Xuanwo) [inline src/meta/protos/proto/table.proto:128]: Store our config in `string` is very easy to break.

Maybe we can use `StageStorage`?

https://github.com/datafuselabs/databend/blob/3b78959ca483cbd62abc322faadd1a0c59b2fead/src/meta/protos/proto/user.proto#L131

Visit @drmingdrmer for advice.

Comment (drmingdrmer) [reply src/meta/protos/proto/table.proto:128]: Right, using a structured type would be better.

File: src/query/storages/fuse/src/pruning/bloom_pruner.rs [hunk 5]
```diff
@@ -158,10 +165,17 @@ mod util {
         bloom_index_col_names: &[String],
         index_location: &Location,
         index_length: u64,
+        storage_params: Option<StorageParams>,
```
Comment (drmingdrmer) [inline src/query/storages/fuse/src/pruning/bloom_pruner.rs:168]: There is already a `dal` arg at line 162, why does it need `storage_params` at all?

I assume the `dal` can get any data it needs, without depending on any additional parameters.

This weirdness may be rooted in the previous design: the `dal` is not used but instead another **reader** is created.

**Pull request closed without merging**